Geometry_Engine/Compute/CurveProximity.cs

```diff
@@ -0,0 +1,549 @@
+            for (int i = 1; i < temp.Count; i++)
+            {
+                cp = curve2.CurveProximity(temp[i]);
+                if (cp.Item1.Distance(cp.Item2) < result.Item1.Distance(result.Item2))
+                {
+                    result = cp;
+                }
+            }
+            return result;
+        }
+
+        /***************************************************/
+
+        public static BH.oM.Reflection.Output<Point, Point> CurveProximity(this Polyline curve1, PolyCurve curve2, double tolerance = Tolerance.Distance)
+        {
+            return curve2.CurveProximity(curve1);
+        }
+
+        /***************************************************/
+
+        public static BH.oM.Reflection.Output<Point, Point> CurveProximity(this Polyline curve1, Polyline curve2, double tolerance = Tolerance.Distance)
+        {
+            List<Line> temp = new List<Line>();
+            for (int i = 0; i < curve1.ControlPoints.Count - 1; i++)
+                temp.Add(Create.Line(curve1.ControlPoints[i], curve1.ControlPoints[i + 1]));
+            BH.oM.Reflection.Output<Point, Point> result = curve2.CurveProximity(temp[0]);
+            BH.oM.Reflection.Output<Point, Point> cp = new BH.oM.Reflection.Output<Point, Point>();
+            for (int i = 1; i < temp.Count; i++)
+            {
+                cp = curve2.CurveProximity(temp[i]);
+                if (cp.Item1.Distance(cp.Item2) < result.Item1.Distance(result.Item2))
+                {
+                    result = cp;
+                }
+            }
+            return result;
+        }
+
+        /***************************************************/
+
+        [NotImplemented]
+        public static BH.oM.Reflection.Output<Point, Point> CurveProximity(this ICurve curve1, Ellipse curve2, double tolerance = Tolerance.Distance)
+        {
+            throw new NotImplementedException();
+        }
+
+        /***************************************************/
+
+        [NotImplemented]
+        public static BH.oM.Reflection.Output<Point, Point> CurveProximity(this ICurve curve1, NurbsCurve curve2, double tolerance = Tolerance.Distance)
+        {
+            throw new NotImplementedException();
+        }
+
+        /***************************************************/
+        /**** Public Methods - Interfaces               ****/
+        /***************************************************/
+
+        public static BH.oM.Reflection.Output<Point,Point> ICurveProximity(this ICurve curve1, ICurve curve2, double tolerance = Tolerance.Distance)
+        {
+            return CurveProximity(curve1 as dynamic, curve2 as dynamic);
+        }
+
+        /***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+
+        internal static double[] SkewLineProximity(this Line line1, Line line2, double angleTolerance = Tolerance.Angle)
```

Comment:
Is there a specific reason why this is `internal`? Why can't it fall into either `public` or `private`?

Reply:
That was my suggestion. I do not see the `public` use of the method (and do not want to flood the UI with useless stuff), but it cannot be `private` as it is used in 2 classes, therefore `internal`. What do you think?

Reply:
I remember we have had a pretty bad experience with `internal` methods.  Usually when you don't see a use case for a method is just because you haven't found it yet (it may be whatever). I vouch for the `public` one in here.

Reply:
What do you mean by pretty bad experience?

I would prefer not to change the return type of the method, which would result in a public method returning an array. Are you fine with that?

Reply:
Mainly about wanting to access the method from other assemblies, and we couldn't. @adecler and @IsakNaslundBh can tell more than me about that.

For the return type, it shouldn't return an array even if it is `private` or `internal` preferably.

As for not exposing the method, take a look at the Reflection_Engine. There are a lot of methods that have very few use cases, some of them has none, but they're `public` anyway.

Reply:
Yeah, the point of using `internal` is to expose the method to other classes inside the assembly  while preventing it from being used by other assemblies, right? Concerning the return type, `SkewLineProximity` has been there for some time, not sure if there is any point in changing it now.

Just to be clear, I generally agree with your points, but in this particular case my idea was to implement as few changes as possible, i.e. not to expose the hidden method and not to change its return type. That's why `internal` and array.

Reply:
I get your point and I still don't see a reason for not exposing it. Exposing even a hundred methods will not flood the UI, since we have more than 5000 methods there already.

In any case, we should not use access modifiers as a tool to hide/expose items from the UI, because it's not their purpose. They are usually used because exposing it publicly harms the logic of the class it is in, making it invalid - which is (almost) never our case with `static` classes.

Few changes possible is an empty diff - you mean to change as little as possible to implement the functionality? If so, compliance is part of that. The method is still not compliant, so at one point it will be changed anyway.

At least this is my opinion, but it's up to @adecler, I can just advice on what I see.

Reply:
The reason for not exposing is that it is a helper method that was sitting there for a year and would still be kept `private` if it was not used in 2 different classes due to the changes made in this PR.

Following your logic, does it mean that any helper method, whatever it does, if used by two classes need to go `public`?

Reply:
No, because there is no such thing as a helper method the way the framework is configured.
Every method should be `public`, unless there is a very valid reason for not doing so, or it sits in a particular part of the diamond (e.g. Adapter, UI).

Again, I haven't require changes to the pr, feel free to do as you please.

Reply:
> No, because there is no such thing as a helper method the way the framework is configured.
> Every method should be `public`, unless there is a very valid reason for not doing so, or it sits in a particular part of the diamond (e.g. Adapter, UI).

Is that documented somewhere? Personally I have seen more than a few `private` helper methods, also created a few myself, never heard any complaints. Just trying to get my head around this.

Reply:
After another thought and a background conversation with @epignatelli I would opt for making `SkewLineProximity` public with `Output<double, double>` return type. However, it remains unclear to me whether the `private` helper methods are not compliant and will be removed/refactored or just not preferable - @al-fisher @rwemay @adecler could you please reflect on that? Thanks!